### PR TITLE
Remove serde_derive dependency, inlining expanded impls into source code

### PR DESCRIPTION
**Pull request by eqrion**
This commit removes the dependency on `serde_derive` so that
cbindgen doesn't depend on a stale package for OS's like debian.

A new file, README.serde_derive is added with instructions on
how to regenerate the impls with changes.

### CODE REVIEW

**Comment by eqrion** [comment]
r? @emilio 

You obviously don't have to review the copy and pasted impl's, I'm more interested in a sanity check.

This will unblock the uploading of `cbindgen` to [Debian](https://github.com/eqrion/cbindgen/issues/203#issuecomment-429603319) and shift the burden of this issue from users to the maintainers.

There might be a CI failure, I didn't run this through rustfmt yet. Hopefully the autogenerated stuff is fine.

**Comment by eqrion** [comment]
... and there are rustfmt errors. I'll sprinkle `#[cfg_attr(rustfmt, rustfmt_skip)]` over the commit and repush.

**Comment by eqrion** [comment]
> Be aware that the generated code uses serde private APIs so is not guaranteed to compile or do the right thing against newer serde versions. That's fine within cbindgen as long as the impls are regenerated every time serde is bumped in Cargo.lock, but means that if you want other crates to use cbindgen as a library then you will need to pin a specific serde version in cbindgen's Cargo.toml. Otherwise things will break (frequently).

That's an understandable policy, I hadn't thought of that.

Thinking about it further, I'm not sure we can even rely on `Cargo.lock` for binaries as `cargo install cbindgen` doesn't use the lock file in my experience.

File: README.serde_derive [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+5. Find the appropriate `impl`s and copy and paste them into the appropriate `mod`s.
+    1. Keep the `impl`s in source file order
+    2. Unindent the `impl`s to be aligned properly
+6. Update the comments with the version of `serde_derive` used to generate the `impl`s
+   1. You can find this in `Cargo.lock`
+7. Revert the changes to `Cargo.toml`
+8. Delete `expanded.rs`
+9. Test your changes
+
+This is very unfortunate, but it's better than keeping a stale dependency on `serde_derive`.
```
Comment (emilio) [inline README.serde_derive]: I think I disagree with this FWIW.

**Comment by eqrion** [comment]
I suppose this change amounts to pinning `serde` instead of pinning `serde_derive`.

There is one benefit I could see to this, we can update the pinned version of `serde` when needed but currently can't with `serde_derive`. Still not a great solution.

File: src/bindgen/cargo/cargo_lock.rs [hunk 2]
```diff
@@ -29,13 +29,17 @@ impl From<toml::de::Error> for Error {
...
+#[derive(Clone, Debug)]
 pub struct Lock {
     pub root: Option<Package>,
     pub package: Option<Vec<Package>>,
 }
 
-#[derive(Clone, Deserialize, Debug)]
+// Warning: Do not change this without regenerating serde_derive impls.
+// See issue #203 and README.serde_derive for more information.
+#[derive(Clone, Debug)]
```
Comment (emilio) [inline src/bindgen/cargo/cargo_lock.rs]: These are all missing `#[cfg_feature]` bits right?

**Comment by emilio** [comment]
Actually bindgen has the same issue (see https://github.com/rust-lang-nursery/rust-bindgen/pull/1409)... Sigh. I'd rather go fix https://github.com/rust-lang/cargo/issues/2589 instead fwiw.

**Comment by dtolnay** [comment]
If the reason that this serde_derive dependency is a problem is that it enables proc-macro2's "proc-macro" feature and that ends up trying to dynamically link the compiler into cbindgen, then https://github.com/alexcrichton/proc-macro2/issues/139#issuecomment-429694118 may be a shorter path to fix that. We'd like for libproc_macro's build story to be no different from libcore or liballoc, meaning no more dynamic dependency on the compiler.

https://github.com/rust-lang/rust/pull/49219 is much of the work to get there.

**Comment by eqrion** [comment]
I'm going to close this PR and put this on hold as it doesn't really sound like an attractive option. The work around libproc-macro sounds like a promising path.

**Pull request closed without merging**